README
cppman is a project that generates C++ man pages from cplusplus.com

cppman, the man page browser/fetcher generates C++ manual pages from cplusplus.com and provide a man-like interface to view man pages.

Arch Linux users can install the AUR package 'cppman'.

* manpages-cpp is renamed to cppman since September 19, 2012

Forked on November 25th, 2012
Now using python3


ChangeLog
------------------------------------------------------------------------------
cppman Changelog
------------------------------------------------------------------------------
cppman 0.2.1 (November 25th, 2012):
    
    Misc:
    * Change to use python3, with 2to3

cppman 2.2.0 (January 19th, 2013):
	
	Bug fixed:
	* Fix formatting due to the change of cplusplus.com
	* Fix syntax highlighting in vim

	Features added:
	* Reimplement table converter, now support unlimited column table
	  generation

cppman 0.2.0 (September 19th, 2012):

	Bug fixed:
	* Fix formatting due to the change of cplusplus.com

	Misc:
	* Rename project from manpages-cpp to cppman

cppman 0.1.9 (April 22th, 2012):

	Bug fixed:
	* Fix formatting due to the change of cplusplus.com

cppman 0.1.8 (November 29th, 2011):

	Bug fixed:
	* Fix formatting due to the change of cplusplus.com


cppman 0.1.7 (October 7th, 2011):

	Bug fixed:
	* Minor formating bug

	Features added:
	* Pager is now configurable, either 'vim' or 'less' is accepted.
	* Integration of mandb is now configurable, default disabled.

cppman 0.1.6 (April 2nd, 2011):

	Bug fixed:
	* Fix formatting due to the change of cplusplus.com

cppman 0.1.5 (January 31th, 2011):

	Bug fixed:
	* Backslashs in EXAMPLE section disapears
	* Empty "#include" line in some pages

	Features added:
	* Syntax highlighting support for SYNOPSIS and EXAMPLE section
	* Update database

cppman 0.1.3 (September 26th, 2010):

	Features added:
	* Add BSD support (Issue #1)

cppman 0.1.2 (September 21th, 2010):

	Bugs fixed:
	* Minor bug fix

cppman 0.1.1 (September 11th, 2010):

	Bugs fixed:
	* Minor bug fix

	Features added:
	* Enable mandb support

cppman 0.1.0 (September 11th, 2010):

	* Initial release


cppman/cppman.py
#-*- coding: utf-8 -*-
# 
# cppman.py 
#
# Copyright (C) 2010 -  Wei-Ning Huang (AZ) <aitjcize@gmail.com>
# All Rights reserved.
#
# This file is part of cppman.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

import gzip
import os
import os.path
import platform
import re
import shutil
import sqlite3
import subprocess
import sys
import urllib.request, urllib.parse, urllib.error

from . import Environ
from . import Formatter
from .Crawler import Crawler

class cppman(Crawler):
    """Manage cpp man pages, indexes"""
    def __init__(self, forced=False):
        Crawler.__init__(self)
        self.forced = forced
        self.success_count = None
        self.failure_count = None

        self.blacklist = [
            'http://www.cplusplus.com/reference/stl/'
        ]
        self.name_exceptions = [
            'http://www.cplusplus.com/reference/string/swap/'
        ]

        # They are headers, set name to std::name else conflict with other
        # classes/templates
        self.std = [
            'http://www.cplusplus.com/reference/std/limits/',
            'http://www.cplusplus.com/reference/std/new/',
            'http://www.cplusplus.com/reference/std/typeinfo/',
            'http://www.cplusplus.com/reference/std/stdexcept/',
            'http://www.cplusplus.com/reference/std/utility/',
            'http://www.cplusplus.com/reference/std/functional/',
            'http://www.cplusplus.com/reference/std/locale/'
        ]
        self.stl = [
            'http://www.cplusplus.com/reference/stl/bitset/',
            'http://www.cplusplus.com/reference/stl/deque/',
            'http://www.cplusplus.com/reference/stl/list/',
            'http://www.cplusplus.com/reference/stl/map/',
            'http://www.cplusplus.com/reference/stl/multimap/',
            'http://www.cplusplus.com/reference/stl/multiset/',
            'http://www.cplusplus.com/reference/stl/priority_queue/',
            'http://www.cplusplus.com/reference/stl/queue/',
            'http://www.cplusplus.com/reference/stl/set/',
            'http://www.cplusplus.com/reference/stl/stack/',
            'http://www.cplusplus.com/reference/stl/vector/',
            'http://www.cplusplus.com/reference/std/iterator/',
            'http://www.cplusplus.com/reference/algorithm/',
            'http://www.cplusplus.com/reference/std/complex/',
            'http://www.cplusplus.com/reference/std/valarray/',
            'http://www.cplusplus.com/reference/std/numeric/'
        ]

    def extract_name(self, data):
        """Extract man page name from cplusplus web page."""
        name = re.search('<h1>(.+?)</h1>', data).group(1)
        name = re.sub(r'<([^>]+)>', r'', name)
        name = re.sub(r'&gt;', r'>', name)
        name = re.sub(r'&lt;', r'<', name)
        return name

    def rebuild_index(self):
        """Rebuild index database from cplusplus.com."""
        try:
            os.remove(Environ.index_db_re)
        except: pass
        self.db_conn = sqlite3.connect(Environ.index_db_re)
        self.db_cursor = self.db_conn.cursor()
        self.db_cursor.execute('CREATE TABLE CPPMAN (name VARCHAR(255), '
                               'url VARCHAR(255))')
        try:
            self.crawl(self.insert_index)

            # Rename dumplicate entries
            dumplicates = self.db_cursor.execute('SELECT name, COUNT(name) '
                                                 'AS NON '
                                                 'FROM CPPMAN '
                                                 'GROUP BY NAME '
                                                 'HAVING (NON > 1)').fetchall()
            for name, num in dumplicates:
                dump = self.db_cursor.execute('SELECT name, url FROM CPPMAN '
                                              'WHERE name="%s"'
                                              % name).fetchall()
                for n, u in dump:
                    if u not in self.name_exceptions:
                        new_name = re.search('/([^/]+)/%s/$' % n, u).group(1)\
                                                                + '::' + n
                        self.db_cursor.execute('UPDATE CPPMAN '
                                               'SET name="%s", url="%s" '
                                               'WHERE url="%s"' %
                                               (new_name, u, u))
            self.db_conn.commit()
        except KeyboardInterrupt:
            os.remove(Environ.index_db_re)
            raise KeyboardInterrupt
        finally:
            self.db_conn.close()

    def insert_index(self, url):
        """callback to insert index"""
        if url not in self.blacklist:
            print("Indexing '%s' ..." % url)
            name = self.extract_name(urllib.request.urlopen(url).read())
            if url in self.std:
                name = 'std::' + name
            elif url in self.stl:
                name = 'stl::' + name
            self.db_cursor.execute('INSERT INTO CPPMAN (name, url) VALUES '
                                   '("%s", "%s")' % (name, url))
        else:
            print("Skipping blacklisted page '%s' ..." % url)

    def cache_all(self):
        """Cache all available man pages from cplusplus.com"""
        print('By defualt, cppman fetch pages on the fly if coressponding '\
            'page is not found in the cache. The "cache-all" option is only '\
            'useful if you want to view man pages offline.')
        print('Caching all contents from cplusplus.com will take about 20 '\
            'minutes, do you want to continue [Y/n]?', end=' ')

        respond = input()
        if respond.lower() not in ['y', 'ye', 'yes']:
            raise KeyboardInterrupt

        try:
            os.makedirs(Environ.man_dir)
        except: pass

        self.success_count = 0
        self.failure_count = 0

        if not os.path.exists(Environ.index_db):
            raise RuntimeError("can't find index.db")

        conn = sqlite3.connect(Environ.index_db)
        cursor = conn.cursor()

        data = cursor.execute('SELECT name, url FROM CPPMAN').fetchall()

        for name, url in data:
            try:
                print('Caching %s ...' % name)
                self.cache_man_page(url, name)
            except Exception as e:
                print('Error caching %s ...', name)
                self.failure_count += 1
            else:
                self.success_count += 1
        conn.close()

        print('\n%d manual pages cahced successfully.' % self.success_count)
        print('%d manual pages failed to cache.' % self.failure_count)
        print('\nrunning mandb...')
        self.update_mandb(False)

    def cache_man_page(self, url, name=None):
        """callback to cache new man page"""
        data = urllib.request.urlopen(url).read()
        groff_text = Formatter.cplusplus2groff(data)
        if not name: name = self.extract_name(data)

        # Skip if already exists, override if forced flag is true
        outname = Environ.man_dir + name + '.3.gz'
        if os.path.exists(outname) and not self.forced:
            return
        f = gzip.open(outname, 'w')
        f.write(groff_text)
        f.close()

    def clear_cache(self):
        """Clear all cache in man3"""
        shutil.rmtree(Environ.man_dir)

    def man(self, pattern):
        """Call viewer.sh to view man page"""
        try:
            os.makedirs(Environ.man_dir)
        except: pass

        avail = os.listdir(Environ.man_dir)

        if not os.path.exists(Environ.index_db):
            raise RuntimeError("can't find index.db")

        conn = sqlite3.connect(Environ.index_db)
        cursor = conn.cursor()

        # Try direct match
        try:
            page_name, url = cursor.execute('SELECT name,url FROM CPPMAN WHERE'
                                ' name="%s"' % pattern).fetchone()
        except TypeError:
            # Try standard library
            try:
                page_name, url = cursor.execute('SELECT name,url FROM CPPMAN'
                                ' WHERE name="std::%s"' % pattern).fetchone()
            except TypeError:
                # Try STL
                try:
                    page_name, url = cursor.execute('SELECT name,url FROM '
                        'CPPMAN WHERE name="stl::%s"' % pattern).fetchone()
                except TypeError:
                    # Try ambiguous search
                    try:
                        page_name, url = cursor.execute('SELECT name,url FROM '
                            'CPPMAN WHERE name LIKE "%%%s%%"' %
                                                        pattern).fetchone()
                    except TypeError:
                        raise RuntimeError('No manual entry for ' + pattern)
        finally:
            conn.close()

        if page_name + '.3.gz' not in avail or self.forced:
            self.cache_man_page(url, page_name)
            self.update_mandb()

        # Call viewer
        pid = os.fork()
        if pid == 0:
            os.execl(Environ.pager, Environ.pager,
                     Environ.man_dir + page_name + '.3.gz',
                     str(Formatter.get_width()), Environ.pager_config)
        return pid

    def find(self, pattern):
        """Find pages in database."""

        if not os.path.exists(Environ.index_db):
            raise RuntimeError("can't find index.db")

        conn = sqlite3.connect(Environ.index_db)
        cursor = conn.cursor()
        selected = cursor.execute('SELECT name,url FROM CPPMAN'
                            ' WHERE name LIKE "%%%s%%"' % pattern).fetchall()
        if selected:
            for name, url in selected:
                if os.isatty(sys.stdout.fileno()):
                    print(name.replace(pattern, '\033[1;31m%s\033[0m' % pattern))
                else:
                    print(name)
        else:
            raise RuntimeError('%s: nothing appropriate.' % pattern)

    def update_mandb(self, quiet=True):
        """Update mandb."""
        if not Environ.config.UpdateManPath:
            return
        cmd = 'mandb %s' % (' -q' if quiet else '')
        handle = subprocess.Popen(cmd, shell=True).wait()


cppman/Formatter.py
#-*- coding: utf-8 -*-
# 
# Formatter.py - format html from cplusplus.com to groff syntax
#
# Copyright (C) 2010 -  Wei-Ning Huang (AZ) <aitjcize@gmail.com>
# All Rights reserved.
#
# This file is part of cppman.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

import datetime
import fcntl
import re
import struct
import subprocess
import sys
import termios
import urllib.request, urllib.parse, urllib.error

# Format replacement RE list
# The '.SE' pseudo macro is described in the function: cplusplus2groff
rps = [
        # Header, Name
        (r'\s*<div id="I_type"[^>]*>(.*?)\s*</div>\s*<h1>(.*?)</h1>\s*'
         r'<div id="I_file"[^>]*>(.*?)</div>\s*<div class="C_prototype"[^>]*>'
         r'(.*?)</div>\s*<div id="I_description"[^>]*>(.*?)</div>',
         r'.TH "\2" 3 "%s" "cplusplus.com" "C++ Programmer\'s Manual"\n'
         r'\n.SH NAME\n\2 - \5\n'
         r'\n.SE\n.SH TYPE\n\1\n'
         r'\n.SE\n.SH SYNOPSIS\n#include \3\n.sp\n\4\n'
         r'\n.SE\n.SH DESCRIPTION\n' % datetime.date.today(), re.S),
        (r'\s*<div id="I_type"[^>]*>(.*?)\s*</div>\s*<h1>(.*?)</h1>\s*'
         r'<div id="I_file"[^>]*>(.*?)</div>\s*<div id="I_description"[^>]*>'
         '(.*?)</div>',
         r'.TH "\2" 3 "%s" "cplusplus.com" "C++ Programmer\'s Manual"\n'
         r'\n.SH NAME\n\2 - \4\n'
         r'\n.SE\n.SH TYPE\n\1\n'
         r'\n.SE\n.SH DESCRIPTION\n' % datetime.date.today(), re.S),
        (r'\s*<div id="I_type"[^>]*>(.*?)\s*</div>\s*<h1>(.*?)</h1>\s*'
         r'<div class="C_prototype"[^>]*>(.*?)</div>\s*'
         r'<div id="I_description"[^>]*>(.*?)</div>',
         r'.TH "\2" 3 "%s" "cplusplus.com" "C++ Programmer\'s Manual"\n'
         r'\n.SH NAME\n\2 - \4\n'
         r'\n.SE\n.SH TYPE\n\1\n'
         r'\n.SE\n.SH SYNOPSIS\n\3\n'
         r'\n.SE\n.SH DESCRIPTION\n' % datetime.date.today(), re.S),
        # Remove empty #include
        (r'#include \n.sp\n', r'', 0),
        # Remove empty sections
        (r'\n.SH (.+?)\n+.SE', r'', 0),
        # Section headers
        (r'.*<h3>(.+?)</h3>', r'\n.SE\n.SH "\1"\n', 0),
        # 'ul' tag
        (r'<ul>', r'\n.in +2n\n.sp\n', 0),
        (r'</ul>', r'\n.in\n', 0),
        # 'li' tag
        (r'<li>(.+?)</li>', r'* \1\n.sp\n', 0),
        # 'code' tag
        (r'<code>', r'\n.in +2n\n.sp\n', 0),
        (r'</code>', r'\n.in\n.sp\n', 0),
        # 'samp' tag
        (r'<samp>(.+?)</samp>', r'\n.nf\n\1\n.fi\n', re.S),
        # 'pre' tag
        (r'<pre\s*>(.+?)</pre\s*>', r'\n.nf\n\1\n.fi\n', re.S),
        # Subsections
        (r'<b>(.+?)</b>:<br>', r'.SS \1\n', 0),
        # Member functions / See Also table
        (r'<dl class="links"><dt>.*?<b>(.*?)</b>.*?</dt><dd>(.*?)'
         r'<span class="typ">(.*?)</span></dd></dl>',
         r'\n.IP "\1(3)"\n\2 \3\n', 0),
        # Snippet
        ## Remove table
        (r'<table class="snippet">', r'', 0),
        ## Remove line numbers
        (r'<td class="rownum">.+?</td>', r'', re.S),
        ## remove td
        (r'<td class="source">(.+?)</td>', r'\1', re.S),
        (r'<td class="output">(.+?)</td>', r'\1', re.S),
        # Table
        (r'<table.*?>', r'.TS\nallbox tab(|);\n', 0),
        (r'</table>', r'\n.TE\n.sp\n', 0),
        ## Table head column
        (r'<tr.*?><th.*?>(.*?)</th>', r'TH{\n\1\nTH}', 0),
        (r'<th colspan="(.+?)">(.*?)</th>', r'|TH{S\1\n\2\nTH}', 0),
        (r'<th.*?>(.*?)</th>', r'|TH{\n\1\nTH}', 0),
        ## Table body column
        (r'<tr.*?><td.*?>((.|\n)*?)</td>', r'T{\n\1\nT}', 0),
        (r'<td.*?>((.|\n)*?)</td>', r'|T{\n\1\nT}', 0),
        (r'</tr>', r'\n', 0),
        # Footer
        (r'<div id="CH_bb">.*$',
         r'\n.SE\n.SH REFERENCE\n'
         r'cplusplus.com, 2000-2010 - All rights reserved.', re.S),
        # C++ version tag
        (r'<div title="(C\+\+..)">', r'.sp 2\n\1\n', 0),
        # 'br' tag
        (r'<br>', r'\n.br\n', 0),
        (r'\n.br\n.br\n', r'\n.sp\n', 0),
        # 'dd' 'dt' tag
        (r'<dt>(.+?)</dt>\s*<dd>(.+?)</dd>', r'.IP "\1"\n\2\n', re.S),
        # Bold
        (r'<strong>(.+?)</strong>', r'\n.B \1\n', 0),
        # -
        (r'-', r'\-', 0),
        # Remove row number in EXAMPLE
        (r'<td class="rownum">.*?</td>', r'', re.S),
        # Any other tags
        (r'<script[^>]*>[^<]*</script>', r'', 0),
        (r'<.*?>', r'', re.S),
        # Misc
        (r'&lt;', r'<', 0), 
        (r'&gt;', r'>', 0),
        (r'&amp;', r'&', 0),
        (r'&nbsp;', r' ', 0),
        ('\x0d([^)])', r'\n.br\n\1', 0),
        (r'>/">', r'', 0),
        (r'/">', r'', 0),
        # Remove empty lines
        (r'\n\s*\n+', r'\n', 0),
        (r'\n\n+', r'\n', 0),
        # Preserve \n" in EXAMPLE
        (r'\\n"', r'\en"', 0),
      ]

def cplusplus2groff(data):
    """Convert HTML text from cplusplus.com to Groff-formated text."""
    # Remove sidebar
    try:
        data = data[data.index('<div class="C_doc">'):]
    except ValueError: pass

    # Replace all
    for rp in rps:
        data = re.compile(rp[0], rp[2]).sub(rp[1], data)

    # Process table header
    # Header is like text block, but use TH instead of T
    # 'TH{SX' : X is the colspan of that column
    for table in re.findall(r'\n\.TS(.+?)\.TE', data, re.S):
        tbl = table
        content = ''
        n_cols = 0
        for th in re.findall(r'(TH{.*?TH}(?:\|TH{.*?TH})+)', table, re.S):
            n_cols = 0
            for col in th.split('|'):
                try:
                    colspan = re.match(r'TH{S([0-9]+?)', col).group(1)
                except AttributeError:
                    content += 'c '
                    n_cols += 1
                else:
                    content += 'c ' + 's ' * (int(colspan) - 1)
                    n_cols += int(colspan)
            content += '\n'

        if n_cols == 2:
            content += 'l lx'
        elif n_cols == 3:
            content += 'l lx l'
        else:
            content += 'l ' * n_cols

        content += '.\n'

        # Insert format header
        tbl = tbl.replace('allbox tab(|);\n', 'allbox tab(|);\n%s' % content)
        # Recover TH{SX -> T{
        tbl = re.sub('TH{S[0-9]+?', 'T{', tbl).replace('TH', 'T')

        data = data.replace(table, tbl)

    # Escape column which starts with dot
    data = re.sub(r'T{\n(\..*?)\nT}', r'T{\n\E \1\nT}', data, flags=re.S)

    # Upper case all section headers
    for st in re.findall(r'.SH .*\n', data):
        data = data.replace(st, st.upper())

    # Add tags to member/inherited member functions
    # e.g. insert -> vector::insert
    #
    # .SE is a pseudo macro I created which means 'SECTION END'
    # The reason I use it is because I need a marker to know where section ends.
    # re.findall find patterns which does not overlap, which means if I do this:
    # secs = re.findall(r'\n\.SH "(.+?)"(.+?)\.SH', data, re.S)
    # re.findall will skip the later .SH tag and thus skip the later section.
    # To fix this, '.SE' is used to mark the end of the section so the next
    # '.SH' can be find by re.findall

    page_type =  re.search(r'\n\.SH TYPE\n(.+?)\n', data)
    if page_type and 'class' in page_type.group(1):
        class_name = re.search(r'\n\.SH NAME\n(?:.*::)?(.+?) ', data).group(1)

        secs = re.findall(r'\n\.SH "(.+?)"(.+?)\.SE', data, re.S)

        for sec, content in secs:
            # Member functions
            if 'MEMBER' in sec and 'INHERITED' not in sec and\
               sec != 'MEMBER TYPES':
                contents = re.sub(r'\n\.IP "([^:]+?)"', r'\n\.IP "%s::\1"'
                                  % class_name, content)
                # Replace (constructor) (destructor)
                contents = re.sub(r'\(constructor\)', r'%s' % class_name,
                                  contents)
                contents = re.sub(r'\(destructor\)', r'~%s' % class_name,
                                  contents)
                data = data.replace(content, contents)
            # Inherited member functions
            elif 'MEMBER' in sec and 'INHERITED' in sec:
                inherit = re.search(r'.+?INHERITED FROM (.+)',
                                    sec).group(1).lower()
                contents = re.sub(r'\n\.IP "(.+)"', r'\n\.IP "%s::\1"'
                                  % inherit, content)
                data = data.replace(content, contents)

    # Remove invalid markings in tables
    for st in re.findall(r'\.TS(.+?)\.TE', data, re.S):
        sts = re.sub(r'\n\...\n', r'\n', st)
        sts = re.sub(r'\n\.B (.+?)\n', r'\1', sts)
        sts = re.sub(r'\n\.', r'\n\\.', sts)
        index = data.index(st)
        data = data[:index] + sts + data[index + len(st):]

    # Remove pseudo macro '.SE'
    data = data.replace('\n.SE', '')

    return data

def groff2man(data):
    """Read groff-formated text and output man pages."""
    width = get_width()

    cmd = 'groff -t -Tascii -m man -rLL=%dn -rLT=%dn' % (width, width)
    handle = subprocess.Popen(cmd, shell=True, stdin=subprocess.PIPE,
                                               stdout=subprocess.PIPE,
                                               stderr=subprocess.PIPE)
    man_text, stderr = handle.communicate(data)
    return man_text

def cplusplus2man(data):
    """Convert HTML text from cplusplus.com to man pages."""
    groff_text = cplusplus2groff(data)
    man_text = groff2man(groff_text)
    return man_text

def get_width():
    """Get terminal width"""
    # Get terminal size
    ws = struct.pack("HHHH", 0, 0, 0, 0)
    ws = fcntl.ioctl(0, termios.TIOCGWINSZ, ws)
    lines, columns, x, y = struct.unpack("HHHH", ws)
    width = columns * 39 / 40
    if width >= columns -2: width = columns -2
    return width

def test():
    """Simple Text"""
    name = input('What manual page do you want? ')
    ifs = urllib.request.urlopen('http://www.cplusplus.com/' + name)
    print(cplusplus2man(ifs.read()), end=' ')
    #with open('test.txt') as ifs:
    #    print cplusplus2man(ifs.read()),

if __name__ == '__main__':
    test()
